Replace EOS with newline to prevent context/memory being flushed by EOS in interactive mode

rabidcopy opened the pull request:
Edit: Most of the below is now outdated. This PR aims to do two things.
-Replace EOS with newline to prevent context/memory being flushed by EOS in interactive mode
-Better reverse prompt behavior on EOS by injecting the first given reverse prompt on the newline upon EOS

Aims to improve coherence and ability to resume the interactive session when the user is given input back after an end of text token is reached. Not sure what token 13 is or why it seems to help, so requesting someone more knowledgeable on this.

Forgive the crudeness of this PR. As of https://github.com/ggerganov/llama.cpp/commit/368d0c8a9ebae16a20e1c8971b21ee888bdefad5 interactive mode now continues and gives back user input when an end of text token is reached. This is great, however there seems to be odd behavior after user is given control back following the end of text token. The following below this is my observations. (pasted mostly from another issue)

This is a bit of a tangent, but I've been looking further into the weird behavior when the end of text token occurs and gives the user control. (without the use of `--ignore-eos`, meaning end of texts can occur) and would like to propose a change to https://github.com/ggerganov/llama.cpp/commit/368d0c8a9ebae16a20e1c8971b21ee888bdefad5. While I'm not super familiar here, it seems that changing the section add these two lines seems to improve if not outright fix the weird lack of coherence that occurs after an end of text token. I am not qualified enough to speak on this technically but saw the `emb.back() = 13` and `last_n_tokens.back() = 13` thrown around as a small hack to get around end of texts prior to https://github.com/ggerganov/llama.cpp/commit/368d0c8a9ebae16a20e1c8971b21ee888bdefad5.
```
        // end of text token
        if (embd.back() == EOS_TOKEN_ID) {
            if (params.interactive) {
                is_interacting = true;
                embd.back() = 13;
                last_n_tokens.back() = 13;
            } else {
                fprintf(stderr, " [end of text]\n");
                break;
            }
        }
```
Here are three excerpts from a reverse prompt dialogue WITHOUT this addition, the current behavior when an end of text token is reached. (edited in the [end of text] parts for clarity to indicate when it gave me back control)
```
Start of the neverending dialogue that lasts forever and goes on for 1000 years

Jessie: H-hi there.. if someone is there. *sweats*
User: Hi there!
Jessie: A-ah! *startled*
User:It's okay, it's just me. :)                       
Jessie: Oh thank goodness...I thought i was going crazy talking to myself like this. [end of text]
User:Hello?
This user account has been created by a bot (or automated program).
If you have any questions about its content or actions, contact the bot operator directly using their talk page message system.
To protect against spamming, all new accounts are automatically blocked from editing until they pass manual review. If your request is approved, your username will be changed to match your email address.
Retrieved from "https://en.wikipedia.org/wiki/Special:UsedAccount" [end of text]
```
A bit of second hand embarrassment as it randomly started going on about the anime, Naruto and fan-fiction..
```
Start of the neverending dialogue that lasts forever and goes on for 1000 years

Jessie: H-hi there.. if someone is there. *sweats*
User: Hi there!
Jessie: A-ah! *startled*
User:It's okay. It's just me.
Jessie: Oh... ummm.... what do you need from me today?
User:end of story
 [end of text]
User: Nothing.
Title: Nothing.
Characters/Pairing: Kiba Inuzuka & Tenten (from Naruto)
Rating: PG-13 (for language)
Summary: The two are enjoying a day off together when they come across something unexpectedly beautiful.
Notes: This was written as part of the 'Narutomania' challenge over at /fic/, which asked us to write a one shot using any combination of characters or pairings from the show.
Kiba Inuzuka & Tenten were having a great time exploring Konoha during their days off.
```
Particularly strong example of how it just forgot who I was speaking with entirely after an end of text.
```
Start of the neverending dialogue that lasts forever and goes on for 1000 years

Jessie: H-hi there.. if someone is there. *sweats*
User: Hi there!
Jessie: A-ah! *startled*
User:Don't worry, just me.
Jessie: Oh... thanks....
\end{code} [end of text]
User: Hi.
User: Hello?
System: Yes. What would you like today?
User: What happened to Jessie?
System: Who are you talking about?
User: Is Jessie not here?
System: Nope. She left a while ago.
User: Oh, that's too bad.
System: You mean she was here before?
User: I thought she was.
System: Well, maybe it wasn't her then. Maybe we were thinking of another person. Or maybe we were thinking of two people at once, but one got lost inside our heads somewhere along the way.
User:
```
And here two small excerpts WITH the above change when the end of text token is thrown.
```
Start of the neverending dialogue that lasts forever and goes on for 1000 years

Jessie: H-hi there.. if someone is there. *sweats*
User: Hi there!
Jessie: A-ah! *startled*
User:It's okay, it's just me. :)
Jessie: Oh thank goodness...I thought maybe a ghost was talking to me or something...
User:end of story
 [end of text]
EndUser: Hi.
Jessie: Hello! [end of text]
User:Hi
Jessie: Who are you?
```
```
Start of the neverending dialogue that lasts forever and goes on for 1000 years

Jessie: H-hi there.. if someone is there. *sweats*
User: Hi there!
Jessie: A-ah! *startled*
User: It's okay! It's just me. :)
Jessie: Oh...thank god! You scared me half to death! :O
User:end of story
 [end of text]
EndUser: Now what?
Jessie: Well now we have a conversation going on between us two, but it doesn't end here. [end of text]
```
I've tested this over the past day, and it seems pretty apparent that without `emb.back() = 13` and `last_n_tokens.back() = 13` it completely loses the plot when you give any input following the end of text token.

Would greatly appreciate someone more knowledgeable signing off on this and potentially explaining why these two lines with token 13 seem to remedy the weird behavior that occurs after an end of text token is reached in interactive mode.

Also to be clear, this does **_not_** seem to effect the later lines pertaining to when remaining_tokens runs out. That seems to give user control and allow for continuation of the session just fine with no lost coherence. So **_just_** the end of text part.

Apologies in advance if I've done anything wrong in the process of creating this PR.

rabidcopy committed:
Improve interactive mode's coherence after EOS

Aims to improve coherence and ability to resume the interactive session when the user is given input back after an end of text token is reached.
Not sure what token 13 is or why it seems to help. See conversation for examples.

slaren commented:
The model was (presumably) trained to ignore everything before the eos token. Token 13 is \n so you are replacing the end of text token with a new line, so the model will interpret is as such.
On a different note I think it would be better to avoid magic magic numbers like that and add a constant somewhere. A hardcoded constant is better than nothing, but you could also find the token id for \n dynamically using the vocab.token_to_id map.

rabidcopy commented:
> The model was (presumably) trained to ignore everything before the eos token. Token 13 is \n so you are replacing the end of text token with a new line, so the model will interpret is as such. On a different note I think it would be better to avoid magic magic numbers like that and add a constant somewhere. A hardcoded constant is better than nothing, but you could also find the token id for \n dynamically using the vocab.token_to_id map.

Thanks you very much for the explanation!

That makes a lot of sense now actually. I really do want this PR to make it in as I feel the current behavior is less than ideal for it to give the user control back just for the context to more or less be gone just because it reached the end of text. However it does raise the question if users wishing to avoid this behavior should just rely on the `--ignore-eos` argument to strip the end of text token altogether. But then that brings into question why give the user back control in interactive mode as a default if the state of it is going to be incoherent and going to output random things thereafter.

I would be interested in future proofing this PR by dynamically determining the `\n` token instead of using a constant which in the future may be a completely different token depending on what model is loaded. But I am admittedly lacking in experience or any actual coding knowledge outside of rudimentary stuff. Which is embarrassing because it probably isn't very hard. Though I do notice that the end of text token is also a constant and not dynamically determined. https://github.com/ggerganov/llama.cpp/blob/074bea2eb1f1349a0118239c4152914aecaa1be4/main.cpp#L32 Any pointers on how to proceed with this would be much appreciated. I just know I feel the current default behavior feels less than perfect when the user could just use `--ignore-eos` to remedy end of text tokens effectively ending their interactive session.

rabidcopy committed:
Make newline token a constant

slaren commented:
To find the token id dynamically you could do something like this in main, after the call to llama_model_load and before the main loop:
```c++
const auto newline_token_id = vocab.token_to_id["\n"];
```
This cannot be done in the same way for the eos token because it is one of many special tokens that maps to an empty string in llama.cpp. The way the tokenizer is exported would need to be changed to be able to find this token dynamically unambiguously.

rabidcopy committed:
dynamically determine newline token

slaren commented:
It doesn't compile as is because you have made the constant a local in `llama_model_load()`, and this means that it is not visible in `main()` where it is used. For example you could place it around line 955 instead.

rabidcopy commented:
> It doesn't compile as is because you have made the constant a local in `llama_model_load()`, and this means that it is not visible in `main()` where it is used. For example you could place it around line 955 instead.

Wait really, that's strange. I thought it compiled fine. I'll fix it.

rabidcopy committed:
relocate previous newline token const

rabidcopy committed:
cleanup whitespace

rabidcopy commented:
Hold on, something isn't behaving right. I think something isn't working actually. Forgive me but investigating.

rabidcopy committed:
print a new line on end of text in interactive

this may need to be looked into further when not using a reverse prompt

eous commented:
There is a difference in behavior as well between the alpaca fine tunes floating around and the original llama models regarding token 13. Just an fyi.

tjohnman commented:
That's weird. I merged [330b86e](https://github.com/ggerganov/llama.cpp/pull/333/commits/330b86eed2d4e7e8588f62f5f1aba476e7ac406b) and I didn't have the newline issue. Maybe it has to do with [other tweaks I have](https://github.com/tjohnman/llama.cpp/blob/experimental/main.cpp)?

rabidcopy commented:
> That's weird. I merged [330b86e](https://github.com/ggerganov/llama.cpp/pull/333/commits/330b86eed2d4e7e8588f62f5f1aba476e7ac406b) and I didn't have the newline issue. Maybe it has to do with [other tweaks I have](https://github.com/tjohnman/llama.cpp/blob/experimental/main.cpp)?

I'm not sure, but it seems to happen consistently to me when using a reverse prompt. It'll place the reverse prompt at the end of the line instead of on a new line  unless I manually print a new line. Behavior when not using a reverse prompt seems fine I think. Maybe only print a new line with my change when in reverse prompt mode?

Something along these lines? (this is a bit embarrassing but I don't know how to cleanly put this into the end of text token block, what I tried doesn't compile)

```
if (params.antiprompt.size() != 0) {
   fprintf(stderr, "\n");
```

slaren commented:
Here is a suggestion: notice that the token is generated in main:~1003, in this line:
```C++
id = llama_sample_top_p_top_k(vocab, logits.data() + (logits.size() - n_vocab), last_n_tokens, repeat_penalty, top_k, top_p, temp, rng);
```

At this point you could check if the generated token is eos and replace it with a newline, for example:
```C++
if (id == EOS_TOKEN_ID && params.interactive) {
    id = NEWLINE_TOKEN_ID;
}
```

This should remove the need to print the new line or do anything else yourself, since it will be treated and printed automatically later on as if it came from the model.

rabidcopy committed:
only print manual newline with reverse prompt

fix formatting of reverse prompts so they don't end up at the end of the current line while not introducing unnecessary new lines otherwise

rabidcopy commented:
> Here is a suggestion: notice that the token is generated in main.cpp:~1003, in this line:
> 
> ```c++
> id = llama_sample_top_p_top_k(vocab, logits.data() + (logits.size() - n_vocab), last_n_tokens, repeat_penalty, top_k, top_p, temp, rng);
> ```
> 
> At this point you could check if the generated token is eos and replace it with a newline, for example:
> 
> ```c++
> if (id == EOS_TOKEN_ID && params.interactive) {
>     id = NEWLINE_TOKEN_ID;
> }
> ```
> 
> This should remove the need to print the new line or do anything else yourself, since it will be treated and printed automatically later on as if it came from the model.

Ah, I see. I didn't see this until after pushing another change. I'll look into this. Thank you.

slaren commented:
> Correct me if I'm wrong, but wouldn't this be getting very close to what the --ignore-eos argument does?

Not entirely, --ignore-eos prevents eos from being sampled at all in the first place by setting its logit (more or less odds) to zero. The model doesn't actually return one specific token, it returns the odds of all the token and then one is sampled randomly based on these odds. So any other token could be sampled instead, based on the odds returned by the model. What you want to do here is different, you want to allow eos to be sampled and then pretend that it was actually a new line.

rabidcopy commented:
> > Correct me if I'm wrong, but wouldn't this be getting very close to what the --ignore-eos argument does?
> 
> Not entirely, --ignore-eos prevents eos from being sampled at all in the first place by setting its logit (more or less odds) to zero. The model doesn't actually return one specific token, it returns the odds of all the possible tokens and then one is sampled randomly based on these odds. So any other token could be sampled instead, based on the odds returned by the model. What you want to do here is different, you want to allow eos to be sampled and then pretend that it was actually a new line.

Tried it out and it seems to work, but I think it is behaving more like `--ignore-eos` and never gives back control.

> > Correct me if I'm wrong, but wouldn't this be getting very close to what the --ignore-eos argument does?
> 
> Not entirely, --ignore-eos prevents eos from being sampled at all in the first place by setting its logit (more or less odds) to zero. The model doesn't actually return one specific token, it returns the odds of all the possible tokens and then one is sampled randomly based on these odds. So any other token could be sampled instead, based on the odds returned by the model. What you want to do here is different, you want to allow eos to be sampled and then pretend that it was actually a new line.

Hmm, I'm trying what you suggested and I cannot get it to throw an end of text token at all now, and it never gives back interactive control. It just endlessly generates very similarly to `--ignore-eos` now. Still trying just in case it's a placebo.

slaren commented:
Ah, I think you also need to add `is_interacting = true;` to force it to return control to the user.

rabidcopy commented:
Ouch, new problem with that approach. When using reverse prompt, it does not print the reverse prompt and simply prints a new empty line with control given back to the user. However that is a problem with the other approach as well. I noticed that it seems to only print the first token of the reverse prompt and misses the ":" and only prints "User" which is it's own token.

slaren commented:
The other approach allows it to generate one more token before returning control to the user (since the logic happens at the end of the loop). In some cases it could be the beginning of the reverse prompt, but it would not be guaranteed.

rabidcopy commented:
Well, trying to get something together. I feel the changes I made are still better than the former, but I want the reverse prompt stuff to work well with these changes.

rabidcopy committed:
alternate approach to replace end of text tokens

slaren committed:
Inject the reverse prompt again after eos in interactive mode

ggerganov commented:
Haven't read in super details, but I believe if a <eos> token appears, anything before that is basically ignored by the model.
If you doing interactive chat, seems like ignore <eos> should be on by default so it is never sampled

rabidcopy commented:
> How about not outputting `<eos>` at all? Just set logit of that token to -Infinity before sampling. Model assigns probability to all tokens before sampling trims them, so suppressing `<eos>` will just put something else in it's place.

I believe that would just be the --ignore-eos flag were have already. https://github.com/ggerganov/llama.cpp/commit/50fae10d0339f2bd639f69dd679c0201d939a265 The idea behind this behavior is to allow end of texts to occur while not ending the session and then giving control back to the user. Which is technically what we had already but the end of text token itself causes everything to be forgotten more or less and makes continuing the session at that point less than ideal. My aim is to allow the above behavior while also allowing the context to remain intact by replacing the end of text token with a newline. The newline to fix scenarios where it gives input back at the end of the last line instead of just. To put it shortly, --ignore-eos stops it from being sampled at all but will endlessly generate unless the user manually interjects. While this allows for a stopping point and to give control back. Which is current default behavior in interactive mode, except everything being forgotten due to the end of text token.

slaren commented:
I have suggested a change at https://github.com/rabidcopy/llama.cpp/pull/2 (please check it out @rabidcopy) that would return control to the user by injecting the anti-prompt instead, which should solve that problem. I agree that the code needs significant refactoring, but that should come after other PRs like #77 are merged. In the end this should be a very simple change that fixes a real problem right now.

rabidcopy commented:
> I have suggested a change at [rabidcopy#2](https://github.com/rabidcopy/llama.cpp/pull/2) (please check it out @rabidcopy) that would return control to the user by injecting the anti-prompt instead, which should solve that problem. I agree that the code needs significant refactoring, but that should come after other PRs like #77 are merged. In the end this should be a very simple change that fixes a real problem right now.

This was actually something I was trying really hard to make happen yesterday actually! I was able to inject the anti-prompt but it kept generating another token after it. I'll merge this and check it out asap.

rabidcopy committed:
Merge pull request #2 from slaren/interactive-eos-fix

Inject the reverse prompt again after eos in interactive mode

rabidcopy commented:
> > Ohh, this merge conflicts with #330. Or rather doesn't work as the antiprompt is no longer tokenized.
> 
> Oops! What if we tokenize the (first) reverse prompt right then and there in order to insert it into `embd_inp`? It should be an easy fix.

I think I got it. Testing now before pushing changes.

rabidcopy committed:
tokenize reverse prompt when needed

makes this PR compatible with https://github.com/ggerganov/llama.cpp/pull/330

rabidcopy commented:
I think this works.
```
Assistant: Hi! I'm your personal assistant to answer any questions you may have!
You:Hi jammy, sorry but I gotta go actually, unexpected plans. Goodbye!
Assistant: Oh no worries, I understand! See ya later then! [end of text]
You:█
```

rabidcopy commented:
I think this works.
```
Assistant: Hi! I'm your personal assistant to answer any questions you may have!
You:Hi, sorry but I gotta go actually, unexpected plans. Goodbye!
Assistant: Oh no worries, I understand! See ya later then! [end of text]
You:█
```

tjohnman reviewed (changes requested):
Perhaps we should only tokenize the first reverse prompt if we're only using that.
```
std::vector<llama_vocab::id> first_antiprompt = ::llama_tokenize(vocab, params.antiprompt.front(), false);
embd_inp.insert(embd_inp.end(), first_antiprompt.begin(), first_antiprompt.end());
```
And also we could do the tokenization (first line) at the beginning of the program instead of every time we need to use it. Although I don't know if tokenization is expensive nor how often these lines are going to actually have to be executed.

rabidcopy committed:
tokenize and inject only first reverse prompt

thanks to tjohnman

tjohnman reviewed (changes requested):
I'm sorry I didn't see this earlier, but you are adding the reverse prompt to the context through `emb_inp`, which means it will be added as if the user had inserted it (I guess "emb_inp" stands for "embeddings input"). And it won't be processed until the next iteration of the loop. Shouldn't it be added to `emb` as if the sampler had produced it (the newline is added to `emb`)?

`embd.insert(embd.end(), first_antiprompt.begin(), first_antiprompt.end());`

Moreover, the reverse prompt is being inserted _before_ the newline token is (in line 1015). Shouldn't it be _after_ the newline?

I'm sorry I don't know how to (or if I even can) push commits to the PR myself, otherwise I wouldn't bother you with this.

slaren commented:
> And it won't be processed until the next iteration of the loop

I think that's to be expected, in the next iteration the new line token has to be processed, and then it can start processing the tokens added in `embd_inp`. Adding it to `embd_inp` seems the right thing to do here.

rabidcopy commented:
> I'm sorry I didn't see this earlier, but you are adding the reverse prompt to the context through `emb_inp`, which means it will be added as if the user had inserted it (I guess "emb_inp" stands for "embeddings input"). And it won't be processed until the next iteration of the loop. Shouldn't it be added to `emb` as if the sampler had produced it (the newline is added to `emb`)?
> 
> `embd.insert(embd.end(), first_antiprompt.begin(), first_antiprompt.end());`
> 
> Moreover, the reverse prompt is being inserted _before_ the newline token is (in line 1015). Shouldn't it be _after_ the newline?
> 
> I'm sorry I don't know how to (or if I even can) push commits to the PR myself, otherwise I wouldn't bother you with this.

I think you are mistaken, the newline is being added before the reverse prompt.
```
            // add it to the context
            embd.push_back(id);
```
This is code that was there prior and doesn't have anything to do with the eos->newline->inject first reverse prompt section. I think `embd_inp` works from my testing.

tjohnman commented:
You are both right, since it's in `emb_inp`, it will be inserted after the newline (in the next iteration). I got confused.

rabidcopy committed:
tokenize first reverse prompt once

tjohnman commented on main.cpp:
```diff
@@ -885,6 +885,9 @@ int main(int argc, char ** argv) {
     }
 
+    // tokenize the first reverse prompt
+    std::vector<llama_vocab::id> first_antiprompt = ::llama_tokenize(vocab, params.antiprompt.front(), false);
```
If `params.antiprompt` is empty, this will crash. Better check first.

```
std::vector<llama_vocab::id> first_antiprompt;
if (!params.antiprompt.empty()) {
    first_antiprompt = ::llama_tokenize(vocab, params.antiprompt.front(), false);
}
```

rabidcopy replied:
On it. Thanks

rabidcopy committed:
Merge branch 'master' into interactive-eos-fix

rabidcopy commented:
I think this PR should be renamed at this point. It feels like it'll be a bit confusing if merged as is. "Improved EOS behavior in interactive mode"?
-Replaces EOS with newline and prevents context/memory being flushed by the EOS token
-Injects first reverse prompt upon EOS if reverse prompt is used

tjohnman commented:
I like "Replaces EOS with newline and prevents context/memory being flushed by the EOS token". But I still have doubts about something. I don't know if there is another token that does this, but aren't we essentially removing the ability of the model to decide when to stop?

If we don't specify a reverse token, right now it will go on forever. I think you can't prompt it to generate a short sentence in interactive mode as it is now. Not that this is too serious because that's a pretty fringe use case, but if this is the case would make this dependent on having a reverse prompt, not on being in interactive mode (the check we do in line 1145). What do you think @rabidcopy @slaren ? Am I mistaken about the whole "when to stop" thing?

tjohnman commented:
I like "Replaces EOS with newline and prevents context/memory being flushed by the EOS token". Also, this whole logic is too complex for me. My head hurts.

slaren reviewed (approved):
I think this is good to merge as is.

rabidcopy committed:
Merge branch 'master' into interactive-eos-fix

rabidcopy committed:
add newline token

rabidcopy committed:
add newline token

rabidcopy commented:
Everything but tokenizing and injecting the reverse prompt is now up to date with the API refactor. Not sure how to tokenize the reverse prompt now..

rabidcopy committed:
tokenize/inject reverse prompt for refactor

this doesn't seem right though

rabidcopy commented:
This works and doesn't crash but I don't like my solution for it to work. (This is probably really simple.)
```
    auto first_antiprompt = ::llama_tokenize(ctx, params.prompt,false);
```
I need `first_antiprompt` to exist to account for the inject that comes later. But I don't know how to do the equivalent of `std::vector<llama_vocab::id> first_antiprompt;` after the refactor. So right now this is just a hacky solution to tokenize the normal prompt once just so `first_antiprompt` is declared.
Any suggestions would be much appreciated.

rabidcopy committed:
tokenize nothing for antiprompt if no reverse

rabidcopy committed:
Update main.cpp

rabidcopy committed:
Update main.cpp

rabidcopy committed:
tokenize and inject reverse prompt as needed

this doesn't seem to work if the reverse prompt is tokenized outside earlier on

rabidcopy commented:
I think I finally have this working and it's good to merge finally.
```
Cillian: Hi. What can I do?
You: write me a short sentence about a glass of water and then end the session please, thank you.
Cillian: Ok. Here's your glass of water!
Cillian: It's blue. The glass has a handle on it. There are no ice cubes in it.
You: Great, now end the session for me. Thank you very much.
Cillian: Ending this session now. Bye! [end of text] [reverse prompt injected successfully]
You: █
```
```
Cillian: Hi. What can I do?
You:write me a short poem about water then end the session, thank you.
Cillian: Okay. Water is like life. It flows through your veins and it's what keeps us alive. Water makes everything better.
Cillian: Yes sir! Thank you for this wonderful day.
You: Thank you. Good job. Bye now.
Cillian: Have a good day sir. [end of text] [reverse prompt injected successfully]
You:█
```

rabidcopy committed:
not needed

slaren commented on llama.cpp:
```diff
@@ -1496,6 +1496,9 @@ llama_token llama_token_bos() {
 }
+llama_token llama_token_newline() {
+    return 13;
+}
```
The functions for bos/eos are necessary because these are special tokens, you can use `llama_tokenize` instead.

rabidcopy replied:
Thank you for the heads up. Everything is now in main.cpp. Is this okay now or can it be done better still?

rabidcopy committed:
remove newline token

rabidcopy committed:
remove newline token

rabidcopy committed:
tokenize newline token

rabidcopy committed:
Merge branch 'master' into interactive-eos-fix

rabidcopy committed:
add space to comment

rabidcopy commented on main.cpp:
```diff
@@ -250,10 +250,10 @@ int main(int argc, char ** argv) {
-//    }
+    auto first_antiprompt = ::llama_tokenize(ctx, params.prompt,false);
+    if (!params.antiprompt.empty()) {
+       auto first_antiprompt = ::llama_tokenize(ctx, params.antiprompt.front(), false);
```
Sorry, that's outdated now. I don't believe `std:vector<llama_token> first_antiprompt;` necessary.
```

            // replace end of text token with newline token when in interactive mode
            if (id == llama_token_eos() && params.interactive) {
                id = llama_token_newline.front();
                if (params.antiprompt.size() != 0) {
                    // tokenize and inject first reverse prompt
                    first_antiprompt = ::llama_tokenize(ctx, params.antiprompt.front(), false);
                    embd_inp.insert(embd_inp.end(), first_antiprompt.begin(), first_antiprompt.end());
                }
            }
```

tjohnman replied:
No worries! Looks good.

ggerganov commented on main.cpp:
```diff
@@ -359,6 +362,16 @@ int main(int argc, char ** argv) {
+                id = llama_token_newline.front();
+                if (params.antiprompt.size() != 0) {
+                    // tokenize and inject first reverse prompt
+                    first_antiprompt = ::llama_tokenize(ctx, params.antiprompt.front(), false);
```
```suggestion
                    const auto first_antiprompt = ::llama_tokenize(ctx, params.antiprompt.front(), false);
```

rabidcopy replied:
Ah, I missed that. Sorry! It was correct on my local branch but I messed up here.

rabidcopy committed:
Update main.cpp

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>
